lte/gateway/python/magma/pipelined/app/apn.py

```diff
@@ -0,0 +1,276 @@
+                to specific APN
+            apn: APN UE is connected to with specified IP addr
+        """
+        #TODO(119vik): same IP is reused for several bearers connected to the 
+        # same APN - take care about duplications
+        parser = self._datapath.ofproto_parser
+        
+        encoded_apn = encode_apn(apn)
+        apn_tagging_actions = self._get_apn_tagging_actions(parser, encoded_apn)
+        flow_adds = []
+
+        # Prepare matchers for both UpLink and DownLink traffic
+        outbound_match = MagmaMatch(
+            direction=Direction.OUT, 
+            ipv4_src=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        inbound_match = MagmaMatch(
+            direction=Direction.IN, 
+            ipv4_dst=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        
+        # Generate OF Add Flow messages for Uplink and DownLink traffic
+        for flw in self._get_flows_for_traffic_direction(
+                flows.get_add_resubmit_next_service_flow_msg,
+                outbound_match, 
+                apn_tagging_actions):
+            flow_adds.append(flw)
+
+        for flw in self._get_flows_for_traffic_direction(
+                flows.get_add_resubmit_next_service_flow_msg, 
+                inbound_match, 
+                apn_tagging_actions):
+            flow_adds.append(flw)
+        
+        self.logger.info("Flows to add {}, datapath {}".format(flow_adds, 
+                                                               self._datapath))
+        
+        # Push messages to OVS through async message hub
+        chan = self._msg_hub.send(flow_adds, self._datapath)
+
+        # Wait and process response
+        return self._wait_for_flow_responses(imsi, flow_adds, chan)    
+    
+    def _get_apn_tagging_actions(self, parser, encoded_apn):
+        """
+        Generate list of set value to register openflow actions.
+
+        Ryu doesn't provide API to set value to double extended register 
+        xxreg<x>. However Ryu provides API to match traffic by the value of
+        double extended register xxreg<x>. As a result we should split 
+        single 16bytes value into 4 x 4bytes values so they can be set to 
+        underlying registers reg<4x>..reg<4x+4> which represent single 
+        double extended register xxreg<x>. More details can be found at 
+        http://man7.org/linux/man-pages/man7/ovs-fields.7.html#REGISTER_FIELDS
+
+        Args:
+            parser: 
+            encoded_apn: 16bytes value which represents APN
+        Returns:
+            List of 4  set value to register actions
+        """
+        encoded_apn_registers = split_apn(encoded_apn)
+        return [
+            parser.NXActionRegLoad2(
+                dst=XXREGISTERS_MAP[APN_TAG_REG][reg_num], 
+                value=int(encoded_apn_registers[reg_num], base=16)
+            ) for reg_num in range(4)]
+    
+    def _get_flows_for_traffic_direction(self, generator, matcher, 
```

Comment:
You should either nix this function and expand it at its callsites, or split it into 2 more semantic ones `_get_add_resubmit_flows(direction)`, `_get_delete_resubmit_flows(direction)` that don't use the `generator` parameter. You have an implicit dependency on the signature of the `generator` functions you pass in which an IDE will not be able to detect and refactor if we want to alter those signatures in the future.